### ISSUE TITLE
Import Error undefined symbol: X509_get_signature_nid after upgrade to 16.2.0

### ISSUE THREAD

**Issue by Ovomaltine85** [pyca/pyopenssl#587]
Hi,
today I upgraded from 16.0.0 to 16.2.0 version of pyopenssl. I used pip for it.

Now when I run the python interpreter and try to import the OpenSSL module, I get the following error:
`/usr/local/lib/python2.7/dist-packages/cryptography/hazmat/bindings/_openssl.so: undefined symbol: X509_get_signature_nid`

My installed openssl version under /usr/local/bin/openssl is OpenSSL 1.1.0c  10 Nov 2016

When I run ldd /usr/local/lib/python2.7/dist-packages/cryptography/hazmat/bindings/_openssl.so:
`	linux-vdso.so.1 (0x00007ffc25a7f000)
	libssl.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f987b796000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f987b39a000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f987b17d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f987add2000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f987abce000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f987bcd7000)`
These are all old libraries of openssl. The current ones are lying in /usr/local/lib (eg. libssl.so.1.1). Could this be the problem? How can cryptography / pyopenssl be linked to the new libs?

Thank you for help!

**Comment by Ovomaltine85** [comment]
Thanks! During my research on this problem I already checked out this options.
In my case the libssl* are located in /usr/local/lib, and even if I manually put
`CFLAGS="-I/usr/local/include" LDFLAGS="-L/usr/local/lib" pip install cryptography`
pyopenssl will always be compiled with an libssl1.0.0, which I can see through:
`ldd /usr/local/lib/python2.7/dist-packages/cryptography/hazmat/bindings/_openssl.so`

Due to this, the error still exists. Do you have a clue why cryptography is always built with old libssl, even if you provide the new libs with LDFLAGS?

**Comment by reaperhulk** [comment]
When compiling C modules Python commonly copies the flags it was compiled with over (to be more "compatible"). You might try installing with full verbosity `pip install -vvvvv cryptography` and see if there are flags that point it directly at your old openssl. Unfortunately these problems are frequently highly distribution and environment specific so I can't be much more help than that. Hopefully a distribution will start shipping with 1.1.0 soon-ish.

**Comment by Ovomaltine85** [comment]
What an unlucky mistake by myself. Everytime I built cryptography pip collected the content from the local cache. Your hint with the `-vvvvv` option pointed me to that fact. So I uninstalled cryptography again with the following command by avoiding the cache:
`CFLAGS="-I/usr/local/include" LDFLAGS="-L/usr/local/lib" pip install -I --no-cache-dir cryptography`
Now the _openssl.so is also linked properly with the correct libssl and cryptography produces no more error when imported as python module.

Thank you so much!
